### PR TITLE
feat: adding lessons for else if to pyramid project

### LINKED ISSUE #54825: Add step to teach else if statements in pyramid project
### Describe the Issue

The js beta review project’s step 2 of build a grade book app claims: “Remember that you learned about conditional statements(if, else if, and else).” But actually the “else if” construct was not introduced in the pyramid generator project, only the “if” and “if else” were shown.

please correct this statement or add a step in the pyramid generator project to demo “else if” as it is not a foregone conclusion that newbies will know about it having learned about “if” and “if else”

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/review-js-fundamentals-by-building-a-gradebook-app/step-2

### Your code

N/A

### Expected behavior

Step 2 of gradebook app should not assume foreknowledge of “else if” statements until a lesson has been added to demo it in the pyramid gen project.

### Screenshots

_No response_

### System

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jdwilkin4**
## Summary

The first review project mentions that else if statements were taught in the pyramid. But that ins't the case, So this PR adds that lesson. 

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #54825

### CODE REVIEW

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md [hunk 1]
```diff
@@ -19,6 +19,18 @@ You should not have an `if` statement.
 assert.notMatch(code, /if\s*\(/);
 ```
 
+You should not have an `else if` statement.
```
Comment (JoyShaheb) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md:22]:
```suggestion
You should not use an `else if` statement.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md [hunk 2]
```diff
@@ -19,6 +19,18 @@ You should not have an `if` statement.
...
+
+```js
+assert.notMatch(code, /else\s+if\s*\(/);
+```
+
+You should not have an `else` statement.
```
Comment (JoyShaheb) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md:28]:
```suggestion
You should not use an `else` statement.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/664599653fcd6e97104f9261.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+title: Step 76
+challengeType: 1
+dashedName: step-76
+---
+
+# --description--
+
+Sometimes you will want to run different code when all of the `if...else if` conditions are `false`. You can do this by adding an `else` block. 
+
+An `else` block will only evaluate if the `if` and `else if` blocks do not evaluate.
```
Comment (JoyShaheb) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/664599653fcd6e97104f9261.md:12]:
```suggestion
An `else` block will only evaluate if the conditions in the `if` and `else if` blocks are not met.
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f530d6e33d159e1bf4947.md [hunk 4]
```diff
@@ -1,21 +1,21 @@
...
-Sometimes you will want to run different code when an `if` condition is false. You can do this by adding an `else` block. An `else` block will only evaluate if the `if` block does not evaluate.
+If your pyramid is not inverted, then you will want to have an `else` block that builds the pyramid in the normal order.
 
-The syntax to add an `else` block is:
+In earlier steps, you learned how to work with `else` statement like this:
 
 ```js
 if (condition) {
-  logic
+  // is condition is true, run this code
```
Comment (JoyShaheb) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f530d6e33d159e1bf4947.md:16]:
```suggestion
  // if condition is true, run this code
```

**Review by JoyShaheb — CHANGES_REQUESTED** [review]
All good Jess ! just a few spelling mistakes here and there. Good work 👍

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md [hunk 5]
```diff
@@ -1,15 +1,15 @@
...
+title: Step 77
 challengeType: 1
-dashedName: step-75
+dashedName: step-77
 ---
 
 # --description--
 
-The text is gone again! Empty strings evaluate to `false`, making them a <dfn>falsy value</dfn>. You will learn more about truthy and falsy values in future projects.
+Now that you understand how `if`, `else if` and `else` statements work, you can remove these statements from your code.
```
[the remark this answers is not in the document]
Comment (jdwilkin4) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md:10]: `Now that you understood` doesn't sound right to me.
The sentence could be changed to `Now that you have practiced working with if...else` statements, you can remove them from your code.`

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md (hunk 1 above)
Comment (jdwilkin4) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md:22]: It looks like other assert statements use similar language when asking campers to remove code.

https://github.com/freeCodeCamp/freeCodeCamp/blob/main/curriculum/challenges/english/07-scientific-computing-with-python/learn-how-to-work-with-numbers-and-strings-by-implementing-the-luhn-algorithm/65687c2fd0fec7ba9fb8af30.md

https://github.com/freeCodeCamp/freeCodeCamp/blob/main/curriculum/challenges/english/07-scientific-computing-with-python/learn-list-comprehension-by-building-a-case-converter-program/657f28a0482132aca51a9212.md

I vote we keep this as is for consistency 👍

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md (hunk 2 above)
Comment (jdwilkin4) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md:28]: https://github.com/freeCodeCamp/freeCodeCamp/pull/54830#discussion_r1605257395

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md [hunk 6]
```diff
@@ -64,6 +76,10 @@ function padRow(rowNumber, rowCount) {
 --fcc-editable-region--
 if ("") {
   console.log("Condition is true");
+} else if(5 < 10) {
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/660f44f10ea40f300b896a5e.md:79]:
```suggestion
} else if (5 < 10) {
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/66458f0a05df478aa627629e.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  rows.push(padRow(i, count));
+}*/
+
+--fcc-editable-region--
+if ("") {
+  console.log("Condition is true");
+}
+--fcc-editable-region--
+
+let result = ""
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/66458f0a05df478aa627629e.md:76]:
```suggestion
let result = "";
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/664599653fcd6e97104f9261.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+// TODO: use a different type of loop
+/*for (let i = 1; i <= count; i++) {
+  rows.push(padRow(i, count));
+}*/
+
+--fcc-editable-region--
+if ("") {
+  console.log("Condition is true");
+} else if(5 < 10) {
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/664599653fcd6e97104f9261.md:67]:
```suggestion
} else if (5 < 10) {
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/664599653fcd6e97104f9261.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  console.log("5 is less than 10");
+}
+--fcc-editable-region--
+
+let result = ""
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/664599653fcd6e97104f9261.md:72]:
```suggestion
let result = "";
```

**Review by lasjorg — CHANGES_REQUESTED** [review]
Formatting nitpicks.

**Review by lasjorg — CHANGES_REQUESTED** [review]
Don't really know where to put this, but the `assert.equal(done, 0);` on line 55 of step 77 should be a `strictEqual` so you can't pass with the string `"0"`

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/66458f0a05df478aa627629e.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+Your `else if` statement should check if `5` is less than `10`.
+
+```js
+assert.match(code, /else\s+if\s*\(\s*5\s*<\s*10\s*\)/);
+```
+
+You should log the string `"5 is less than 10"` to the console.
+
+```js
+assert.match(code, /console\.log\(\s*["']5\s+is\s+less\s+than\s+10["']\s*\)/);
```
[the remark this answers is not in the document]
Comment (naomi-lgbt) [reply curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-introductory-javascript-by-building-a-pyramid-generator/66458f0a05df478aa627629e.md:49]:
```suggestion
assert.match(code, /console\.log\(\s*('|"|`)5\s+is\s+less\s+than\s+10\1\s*\)/);
```

We should also be using capture groups to assert the quotes match.

**Comment by lasjorg** [comment]
It did just dawn on me that switching the order of teaching `else` and `else if` might be an idea we should consider.

So teach `if/else` first, then introduce `else if` and put it in the middle of the `if/else` code from the previous step.

First `if/else` step

```
if (false) {
  console.log("If condition is true.");
} else {
  console.log("If condition is false.");
}
```

Then `else if` step

```
if (false) {
  console.log("If condition is true.");
} else if (true) {
  console.log("If condition is true.");
} else {
  console.log("If condition is false.");
}
```

That makes it more clear that `else` can be used with both `if` and `else if`

---

Anyway, I can approve this and we can revisit it later if needed.

**Review by lasjorg — APPROVED** [review]
LGTM, but I think switching the order is something to consider.

**Review by Ksound22 — APPROVED** [review]
LGTM 🚀

**Comment by lasjorg** [comment]
I think the text makes it sound like multiple if statements are the same as a `if/else if`. It doesn't really convey how the condition of the `if` statement the `else if` is connected relates. If you look at how MDN explains it they have a nested example.

```
if (condition1)
  statement1
else
  if (condition2)
    statement2
  else
    if (condition3)
      statement3
```

Maybe we can nest the new `if` statement inside the `else` block in the seed code and then have them move it out to be a `else if`

Seed code:

```
if ("") {
  console.log("Condition is true");
} else {
  if (5 < 10) {
    console.log("5 is less than 10");
  }
}
```

Solution:

```
if ("") {
  console.log("Condition is true");
} else if (5 < 10) {
  console.log("5 is less than 10");
}
```

---

This needs to be changed as well somehow.

> But instead of using multiple `if` statements, you can use `else if` statements to check multiple conditions in a single block of code.

Maybe something like:

> Instead of nesting `if` and `else` statements, you can use `else if` statements.

**Comment by jdwilkin4** [comment]
I have reverted the changes back the way it was before the if, else if switch.
I think it would be better to iron this approach out more because anytime we ask campers to move code around, it needs to be super clear in the seed code where to move it. And the hint text needs to be super clear too.

I vote we merge this in as is.
Then open a separate issue to hash out the details for the if else if stuff.
Then a new PR can be opened to iterate off of this

**Comment by lasjorg** [comment]
Sure, we can do that.

Anyway, here is one way we might write the instructions.

> Move the nested `if` statement out of the `else` block and combine the two so it creates a `else if` block. You can look at the example code for help with the syntax.

**Review by lasjorg — APPROVED** [review]
Didn't test it, but if this just reverts to what it was before, then I already approved that.

**Review by JoyShaheb — APPROVED** [review]

**Review by naomi-lgbt — APPROVED** [review]

**Pull request merged**